## Problème 76

It is possible to write five as a sum in exactly six different ways:

```
4 + 1
3 + 2
3 + 1 + 1
2 + 2 + 1
2 + 1 + 1 + 1
1 + 1 + 1 + 1 + 1
```

How many different ways can one hundred be written as a sum of at least two positive integers?

On va compter `M[i,j]` le nombre de décompositions de $i$ en utilisant des entiers entre $1$ et $j$. Ultimement on veut `M[100,99]`.

In [1]:
def Pb76():
    MAX = 100

    M = np.zeros((MAX+1,MAX+1),dtype=int)

    for i in range(1,MAX+1):
        M[i,1]=1
        M[1,i]=1

    for i in range(1,MAX+1):
        for j in range(2,i):
            M[i,j]=M[i,j-1]+M[i-j,j]
        M[i,i]=M[i,i-1]+1
        for j in range(i+1,MAX+1):
            M[i,j]=M[i,i]
            
    return M[100,99]

## Problème 77

It is possible to write ten as the sum of primes in exactly five different ways:

```
7 + 3
5 + 5
5 + 3 + 2
3 + 3 + 2 + 2
2 + 2 + 2 + 2 + 2
```

What is the first value which can be written as the sum of primes in over five thousand different ways?

On va compter `M[i,j]` le nombre de décompositions de $i$ en utilisant des entiers premiers entre $1$ et $j$.

In [2]:
from sympy import primerange

def Pb77():
    
    def count(n):
        MAX = n

        PRIME = set(primerange(1,MAX+1))

        M = [[0 for _ in range(MAX+1)] for _ in range(MAX+1)]

        for i in tqdm(range(2,MAX+1)):

            for j in range(2,i):
                M[i][j]= M[i][j-1] + ( M[i-j][j] if (j in PRIME) else 0 )

            M[i][i]=M[i][i-1] + (1 if (i in PRIME) else 0)

            for j in range(i+1,MAX+1):
                M[i][j]=M[i][i]

        for i in range(2,MAX+1):
            if M[i][i]>5000:
                return i

        return False

    count(100)

## Problème 78

Let p(n) represent the number of different ways in which n coins can be separated into piles. For example, five coins can be separated into piles in exactly seven different ways, so p(5)=7.

```
OOOOO
OOOO   O
OOO   OO
OOO   O   O
OO   OO   O
OO   O   O   O
O   O   O   O   O
```

Find the least value of n for which p(n) is divisible by one million.

In [3]:
from sympy import partition

def Pb78():
    k=1
    m=10**6
    while partition(k)%m:
        k+=1
    return k

## Problème 79

A common security method used for online banking is to ask the user for three random characters from a passcode. For example, if the passcode was $531278$, they may ask for the 2nd, 3rd, and 5th characters; the expected reply would be: $317$.

The text file, keylog.txt, contains fifty successful login attempts.

Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of unknown length.


In [4]:
class graph:

    def __init__(self,inE,outE):
        self.inE = inE
        self.outE = outE

    def new():
        return graph({},{})

    def add_vertex(self,v):
        if v not in self.inE:
            self.inE[v]=set()
            self.outE[v]=set()
        return None

    def transpose(self):
        self.inE,self.outE=self.outE,self.inE
        return None

    def prune_vertex(self,v):
        if v in self.inE:
            for w in self.inE[v]:
                self.outE[w].remove(v)
            for w in self.outE[v]:
                self.inE[w].remove(v)
            del self.inE[v]
            del self.outE[v]
        return None

    def vertex_list(self):
        return list(self.inE)

    def size(self):
        return len(self.inE)

    def add_edge(self,v,w):
        self.outE[v].add(w)
        self.inE[w].add(v)
        return None

    def successors(self,v):
        return self.outE[v]

    def predecessors(self,v):
        return self.inE[v]

    def in_degree(self,v):
        return len(self.inE[v])
        
    def out_degree(self,v):
        return len(self.outE[v])

    def degree(self,v):
        return out_degree(self,v)+in_degree(self,v)

In [5]:
def Pb79():
    L = [129, 160, 162, 168, 180, 289, 290, 316, 318, 319, 362, 368, 380, 389, 620, 629, 680, 689, 690, 710, 716, 718, 719, 720, 728, 729, 731, 736, 760, 762, 769, 790, 890]
    S = [str(i) for i in L]
    G = graph.new()

    for tri in S:
        for c in tri:
            G.add_vertex(c)
        for i in range(2):
            G.add_edge(tri[i],tri[i+1])
            
    WORD=[]
            
    while G.size()>0:
        for c in G.inE:
            if G.inE[c] == set():
                WORD.append(c)
                G.prune_vertex(c)
                break
    
    return int(''.join(WORD))

## Problème 80

It is well known that if the square root of a natural number is not an integer, then it is irrational. The decimal expansion of such square roots is infinite without any repeating pattern at all.

The square root of two is $1.41421356237309504880..$, and the digital sum of the first one hundred decimal digits is $475$.

For the first one hundred natural numbers, find the total of the digital sums of the first one hundred decimal digits for all the irrational square roots.

In [6]:
from decimal import *

def Pb80():
    getcontext().prec = 110
    S=0

    L=set(range(1,101))
    for i in range(1,11):
        L.remove(i*i)
    
    for i in L:
        for c in str(Decimal(i).sqrt()).replace('.','')[:100] :
            S+=int(c)
    return S

## Problème 81

In [7]:
import numpy as np

def Pb81():
    with open('files/Pb81.txt','r') as file:
        Lines=file.read().split('\n')[:-1]
    Grid=[line.split(',') for line in Lines]

    SIZE = len(Grid)

    for i in range(SIZE):
        for j in range(SIZE):
            Grid[i][j] = int(S[i][j])

    Mat=np.matrix(Grid)
    Sums=np.zeros((SIZE,SIZE),dtype=int)

    Sums[0,0]=Mat[0,0]
    for i in range(1,SIZE):
        Sums[i,0]=Sums[i-1,0]+Mat[i,0]
        Sums[0,i]=Sums[0,i-1]+Mat[0,i]

    for i in range(1,SIZE):
        for j in range(1,SIZE):
            Sums[i,j]=min(Sums[i-1,j],Sums[i,j-1])+Mat[i,j]

    return Sums[SIZE-1,SIZE-1]

## Problème 82

In [8]:
from tqdm import tqdm

def Pb82():
    with open('files/Pb82.txt','r') as file:
        Lines=file.read().split('\n')[:-1]
    Grid=[line.split(',') for line in Lines]
    SIZE = len(Grid)
    for i in range(SIZE):
        for j in range(SIZE):
            Grid[i][j] = int(S[i][j])
    Mat=np.matrix(Grid)
    
    Sums=np.asmatrix(np.zeros((SIZE,SIZE),dtype=int))
    
    for i in range(SIZE):
        Sums[i,0]=Mat[i,0]
                
    for j in tqdm(range(1,SIZE)):
        I = list(range(SIZE))
        TEMP = Sums[:,j-1]+Mat[:,j]
        while len(I)>0:
            i = I[np.argmin(TEMP[I])]
            Sums[i,j]=TEMP[i]
            if i<SIZE-1:
                TEMP[i+1]=min(TEMP[i+1],TEMP[i]+Mat[i+1,j])
            if i>0:
                TEMP[i-1]=min(TEMP[i-1],TEMP[i]+Mat[i-1,j])
            I.remove(i)
        
    return min(Sums[:,SIZE-1])[0,0]

## Problème 83

In [9]:
def Pb83():
    
    with open('files/Pb83.txt','r') as file:
        Lines=file.read().split('\n')[:-1]
    Grid=[line.split(',') for line in Lines]
    SIZE = len(Grid)
    for i in range(SIZE):
        for j in range(SIZE):
            Grid[i][j] = int(S[i][j])
    Mat=np.matrix(Grid)
    
    TEMP=np.asmatrix(np.zeros((SIZE,SIZE),dtype=int))
    TEMP[0,0]=Mat[0,0]
    for i in range(1,SIZE):
        TEMP[i,0]=TEMP[i-1,0]+Mat[i,0]
        TEMP[0,i]=TEMP[0,i-1]+Mat[0,i]
    for i in range(1,SIZE):
        for j in range(1,SIZE):
            TEMP[i,j]=min(TEMP[i-1,j],TEMP[i,j-1])+Mat[i,j]

    VISITED = [(0,0)]
    NEIGH   = [(1,0),(0,1)]
    
    while (SIZE-1,SIZE-1) not in VISITED:
        
        Imin,Jmin = NEIGH[0]
        MIN = TEMP[Imin,Jmin]
        
        for (i,j) in NEIGH[1:]:
            if TEMP[i,j]<MIN:
                Imin=i
                Jmin=j
                MIN=TEMP[i,j]
        
        VISITED.append((Imin,Jmin))
        NEIGH.remove((Imin,Jmin))
        
        if Imin>0:
            TEMP[Imin-1,Jmin]=min(TEMP[Imin-1,Jmin],TEMP[Imin,Jmin]+Mat[Imin-1,Jmin])
            if not ( (Imin-1,Jmin) in NEIGH or (Imin-1,Jmin) in VISITED ) :
                NEIGH.append((Imin-1,Jmin))
        if Imin<SIZE-1:
            TEMP[Imin+1,Jmin]=min(TEMP[Imin+1,Jmin],TEMP[Imin,Jmin]+Mat[Imin+1,Jmin])
            if not ( (Imin+1,Jmin) in NEIGH or (Imin+1,Jmin) in VISITED ) :
                NEIGH.append((Imin+1,Jmin))
        if Jmin>0:
            TEMP[Imin,Jmin-1]=min(TEMP[Imin,Jmin-1],TEMP[Imin,Jmin]+Mat[Imin,Jmin-1])
            if not ( (Imin,Jmin-1) in NEIGH or (Imin,Jmin-1) in VISITED ) :
                NEIGH.append((Imin,Jmin-1))
        if Jmin<SIZE-1:
            TEMP[Imin,Jmin+1]=min(TEMP[Imin,Jmin+1],TEMP[Imin,Jmin]+Mat[Imin,Jmin+1])
            if not ( (Imin,Jmin+1) in NEIGH or (Imin,Jmin+1) in VISITED ) :
                NEIGH.append((Imin,Jmin+1))
        
    return TEMP[SIZE-1,SIZE-1]

## Problème 84

A player starts on the GO square and adds the scores on two 6-sided dice to determine the number of squares they advance in a clockwise direction. Without any further rules we would expect to visit each square with equal probability: 2.5%. However, landing on G2J (Go To Jail), CC (community chest), and CH (chance) changes this distribution.

In addition to G2J, and one card from each of CC and CH, that orders the player to go directly to jail, if a player rolls three consecutive doubles, they do not advance the result of their 3rd roll. Instead they proceed directly to jail.

At the beginning of the game, the CC and CH cards are shuffled. When a player lands on CC or CH they take a card from the top of the respective pile and, after following the instructions, it is returned to the bottom of the pile. There are sixteen cards in each pile, but for the purpose of this problem we are only concerned with cards that order a movement; any instruction not concerned with movement will be ignored and the player will remain on the CC/CH square.

- Community Chest (2/16 cards):
    - Advance to GO
    - Go to JAIL
- Chance (10/16 cards):
    - Advance to GO
    - Go to JAIL
    - Go to C1
    - Go to E3
    - Go to H2
    - Go to R1
    - Go to next R (railway company)
    - Go to next R
    - Go to next U (utility company)
    - Go back 3 squares.

The heart of this problem concerns the likelihood of visiting a particular square. That is, the probability of finishing at that square after a roll. For this reason it should be clear that, with the exception of G2J for which the probability of finishing on it is zero, the CH squares will have the lowest probabilities, as 5/8 request a movement to another square, and it is the final square that the player finishes at on each roll that we are interested in. We shall make no distinction between "Just Visiting" and being sent to JAIL, and we shall also ignore the rule about requiring a double to "get out of jail", assuming that they pay to get out on their next turn.

By starting at GO and numbering the squares sequentially from 00 to 39 we can concatenate these two-digit numbers to produce strings that correspond with sets of squares.

Statistically it can be shown that the three most popular squares, in order, are JAIL (6.24%) = Square 10, E3 (3.18%) = Square 24, and GO (3.09%) = Square 00. So these three most popular squares can be listed with the six-digit modal string: 102400.

If, instead of using two 6-sided dice, two 4-sided dice are used, find the six-digit modal string.

In [10]:
from random import randint
import numpy as np
import matplotlib.pyplot as plt

def Pb84():
    def twodice(n=6):
        return randint(1,n)+randint(1,n)

    CASES = ['GO','A1','CC1','A2','T1','R1','B1','CH1','B2','B3','JAIL',
             'C1','U1','C2','C3','R2','D1','CC2','D2','D3','FP','E1','CH2','E2','E3','R3',
             'F1','F2','U2','F3','G2J','G1','G2','CC3','G3','R4','CH3','H1','T2','H2']
    T = len(CASES)
    CODE={CASES[i]:i for i in range(T)}

    def sample(horizon=10**4,x=0):
        TIME = np.zeros(T)

        for _ in range(horizon):
            x+=twodice(4)
            x%=T

            if CASES[x] == 'G2J':
                x=CODE['JAIL']
            elif CASES[x] in ['CC1','CC2','CC3']:
                match randint(1,16):
                    case 1:
                        x=CODE['GO']
                    case 2:
                        x=CODE['JAIL']
            elif CASES[x] in ['CH1','CH2','CH3']:
                match randint(1,16):
                    case 1:
                        x=CODE['GO']
                    case 2:
                        x=CODE['JAIL']                
                    case 3:
                        x=CODE['C1']
                    case 4:
                        x=CODE['E3']
                    case 5:
                        x=CODE['H2']
                    case 6:
                        x=CODE['R1']                
                    case 7:
                        if CASES[x]=='CH1':
                            x=CODE['R2']
                        elif CASES[x]=='CH2':
                            x=CODE['R3']
                        elif CASES[x]=='CH3':
                            x=CODE['R1']
                    case 8:
                        if CASES[x]=='CH1':
                            x=CODE['R2']
                        elif CASES[x]=='CH2':
                            x=CODE['R3']
                        elif CASES[x]=='CH3':
                            x=CODE['R1']
                    case 9:
                        if CASES[x]=='CH1':
                            x=CODE['U1']
                        elif CASES[x]=='CH2':
                            x=CODE['U2']
                        elif CASES[x]=='CH3':
                            x=CODE['U1']
                    case 10:
                        x-=3
            TIME[x]+=1

        return TIME/horizon

    TIME = np.zeros(T)
    for x0 in range(40):
        TIME+=sample(x=x0)

    i1 = np.argmax(TIME)
    TIME[i1]=0
    i2 = np.argmax(TIME)
    TIME[i2]=0
    i3 = np.argmax(TIME)

    return i1,i2,i3

## Problème 85

In [11]:
from math import sqrt,ceil,floor
from numba import njit

def Pb85():
    
    @njit
    def PROD(i,j):
        return i*(i+1)*j*(j+1)
    
    @njit
    def J(i):
        j = sqrt(4*N / (i*(i+1)))
        return floor(j),ceil(j)
    
    N=2*10**6
    
    ERROR = N
    AREA = 0
    
    for i in range(1,ceil(sqrt(2*sqrt(N)))+1):
    
        jm,jp = J(i)
        
        if abs(4*N-PROD(i,jm))<ERROR:
            ERROR = abs(4*N-PROD(i,jm))
            AREA = i*jm
        if abs(4*N-PROD(i,jp))<ERROR:
            ERROR = abs(4*N-PROD(i,jp))
            AREA = i*jp
    
    return AREA

## Problème 86

In [12]:
from math import isqrt
from numba import njit

def Pb86():
    
    @njit
    def is_square(i):
        return ( i == isqrt(i) ** 2 )
    
    @njit
    def side(M):
        MM=M*M
        T=0
        for bpc in range(2,2*M+1):
            if is_square(bpc**2+MM):
                T += min(M,bpc-1)-(bpc+1)//2+1
        return T

    S=0
    M=0
    while S<10**6:
        M+=1
        S+=side(M)
    return M

## Problème 87

The smallest number expressible as the sum of a prime square, prime cube, and prime fourth power is 28. In fact, there are exactly four numbers below fifty that can be expressed in such a way:


```
28 = 2^2 + 2^3 + 2^4
33 = 3^2 + 2^3 + 2^4
49 = 5^2 + 2^3 + 2^4
47 = 2^2 + 3^3 + 2^4
```

How many numbers below fifty million can be expressed as the sum of a prime square, prime cube, and prime fourth power?

In [13]:
from math import isqrt
from sympy import primerange
from tqdm import tqdm

def Pb87(N=50*10**6):
    
    D={}
   
    A=isqrt(N)+1

    for a in tqdm(primerange(1,A+1),total=len([i for i in primerange(1,A+1)])):
        a2 = a**2
        for b in primerange(1,A+1):
            b3=b**3
            if a2+b3>N:
                break
            for c in primerange(1,A+1):
                tri=a2+b3+c**4
                if tri>N:
                    break
                if tri not in D:
                    D[tri]=1
                else:
                    D[tri]+=1

    return len(D)

## Problème 88

A natural number, N, that can be written as the sum and product of a given set of at least two natural numbers, $\{a_1, a_2, ... , a_k\}$ is called a product-sum number: $N = a_1 + a_2 + ... + a_k = a_1 × a_2 × ... × a_k$.

For example, $6 = 1 + 2 + 3 = 1 × 2 × 3$.

For a given set of size, k, we shall call the smallest N with this property a minimal product-sum number. The minimal product-sum numbers for sets of size, k = 2, 3, 4, 5, and 6 are as follows.

$k=2: 4 = 2 × 2 = 2 + 2$

$k=3: 6 = 1 × 2 × 3 = 1 + 2 + 3$

$k=4: 8 = 1 × 1 × 2 × 4 = 1 + 1 + 2 + 4$

$k=5: 8 = 1 × 1 × 2 × 2 × 2 = 1 + 1 + 2 + 2 + 2$

$k=6: 12 = 1 × 1 × 1 × 1 × 2 × 6 = 1 + 1 + 1 + 1 + 2 + 6$

Hence for $2≤k≤6$, the sum of all the minimal product-sum numbers is $4+6+8+12 = 30$; note that 8 is only counted once in the sum.

In fact, as the complete set of minimal product-sum numbers for $2≤k≤12$ is $\{4, 6, 8, 12, 15, 16\}$, the sum is 61.

What is the sum of all the minimal product-sum numbers for $2≤k≤12000$?

In [14]:
def Pb88(n=12000):

    def FACTORLIST(n):
        L=[[] for _ in range(n+1)]
        L[1].append([])
        for i in range(2,n+1):    
            for j in range(1,n//i + 1):
                p=i*j
                for F in L[j]:
                    L[p].append(F+[i])
        return L
    
    L=FACTORLIST(2*n+1)
    
    MINPRODSUM=[0 for _ in range(n+1)]
    
    for k in range(2,2*n+1):
        for F in L[k]:
            no = k - sum(F)
            psl=len(F)+no
            if psl<=n:
                if MINPRODSUM[psl]:
                    MINPRODSUM[psl]=min(MINPRODSUM[psl],k)
                else:
                    MINPRODSUM[psl]=k
    
    return sum(set(MINPRODSUM[2:n+1]))

## Problème 89

For a number written in Roman numerals to be considered valid there are basic rules which must be followed. Even though the rules allow some numbers to be expressed in more than one way there is always a "best" way of writing a particular number.

For example, it would appear that there are at least six ways of writing the number sixteen:

```
IIIIIIIIIIIIIIII
VIIIIIIIIIII
VVIIIIII
XIIIIII
VVVI
XVI
```

However, according to the rules only XIIIIII and XVI are valid, and the last example is considered to be the most efficient, as it uses the least number of numerals.

The 11K text file, roman.txt (right click and 'Save Link/Target As...'), contains one thousand numbers written in valid, but not necessarily minimal, Roman numerals; see About... Roman Numerals for the definitive rules for this problem.

Find the number of characters saved by writing each of these in their minimal form.

Note: You can assume that all the Roman numerals in the file contain no more than four consecutive identical units.

In [15]:
def compute(S):
    if len(S)==0:
        return 0
    elif S[0]=='M':
        return 1000+compute(S[1:])
    elif len(S)>=2 and S[:2]=='CM':
        return 900+compute(S[2:])
    elif len(S)>=2 and S[:2]=='CD':
        return 400+compute(S[2:])
    elif S[0]=='D':
        return 500+compute(S[1:])
    elif S[0]=='C':
        return 100+compute(S[1:])
    elif len(S)>=2 and S[:2]=='XC':
        return 90+compute(S[2:])
    elif len(S)>=2 and S[:2]=='XL':
        return 40+compute(S[2:])
    elif S[0]=='L':
        return 50+compute(S[1:])
    elif S[0]=='X':
        return 10+compute(S[1:])
    elif len(S)>=2 and S[:2]=='IX':
        return 9+compute(S[2:])
    elif len(S)>=2 and S[:2]=='IV':
        return 4+compute(S[2:])
    elif S[0]=='V':
        return 5+compute(S[1:])
    elif S[0]=='I':
        return 1+compute(S[1:])

def optiter(X):
    X = X.replace('CM','DCCCC')
    X = X.replace('XC','LXXXX')
    X = X.replace('IX','VIIII')
    X = X.replace('CD','CCCC')
    X = X.replace('XL','XXXX')
    X = X.replace('IV','IIII')

    def optimise(X):
        if 'DD' in X:
            return X.replace('DD','M',1)
        elif 'CCCCC' in X:
            return X.replace('CCCCC','D',1)
        elif 'LL' in X:
            return X.replace('LL','C',1)
        elif 'XXXXX' in X:
            return X.replace('XXXXX','L',1)
        elif 'VV' in X:
            return X.replace('VV','X',1)
        elif 'IIIII' in X:
            return X.replace('IIIII','V',1)
        else:
            return X

    Y = optimise(X)
    while X!=Y:
        X,Y=Y,optimise(Y)

    X = X.replace('DCCCC','CM',1)
    X = X.replace('LXXXX','XC',1)
    X = X.replace('VIIII','IX',1)
    X = X.replace('CCCC','CD',1)
    X = X.replace('XXXX','XL',1)
    X = X.replace('IIII','IV',1)

    return X

def Pb89():
    with open('files/Pb89.txt','r') as file:
        Lines=file.read().split('\n')
    return sum([len(c)-len(optiter(c)) for c in Lines])

## Problème 90



Each of the six faces on a cube has a different digit (0 to 9) written on it; the same is done to a second cube. By placing the two cubes side-by-side in different positions we can form a variety of 2-digit numbers.

For example, the square number 64 could be formed:

In fact, by carefully choosing the digits on both cubes it is possible to display all of the square numbers below one-hundred: 01, 04, 09, 16, 25, 36, 49, 64, and 81.

For example, one way this can be achieved is by placing {0, 5, 6, 7, 8, 9} on one cube and {1, 2, 3, 4, 8, 9} on the other cube.

However, for this problem we shall allow the 6 or 9 to be turned upside-down so that an arrangement like {0, 5, 6, 7, 8, 9} and {1, 2, 3, 4, 6, 7} allows for all nine square numbers to be displayed; otherwise it would be impossible to obtain 09.

In determining a distinct arrangement we are interested in the digits on each cube, not the order.
```
{1, 2, 3, 4, 5, 6} is equivalent to {3, 6, 4, 1, 2, 5}
{1, 2, 3, 4, 5, 6} is distinct from {1, 2, 3, 4, 5, 9}
```
But because we are allowing 6 and 9 to be reversed, the two distinct sets in the last example both represent the extended set {1, 2, 3, 4, 5, 6, 9} for the purpose of forming 2-digit numbers.

How many distinct arrangements of the two cubes allow for all of the square numbers to be displayed?


In [16]:
def Pb90():
    
    SQUARE_LIST=['01', '04', '06', '16', '25', '36', '46', '64', '81']

    def satisfy(D1,D2):
        PAIRS=[]
        for a in D1:
            for b in D2:
                PAIRS.append(a+b)
                PAIRS.append(b+a)
        for sq in SQUARE_LIST:
            if sq not in PAIRS:
                return False
        return True

    def dice(L):
        D=set()
        for i in L:
            if i==9:
                D.add('6')
            else:
                D.add(str(i))
        return ''.join(sorted(D))

    DICE_LIST=[]
    for d1 in range(10):
        for d2 in range(d1):
            for d3 in range(d2):
                for d4 in range(d3):
                    for d5 in range(d4):
                        for d6 in range(d5):
                            DICE_LIST.append( dice([d1,d2,d3,d4,d5,d6]) )
    total=0
    for i1 in range(len(DICE_LIST)):
        for i2 in range(i1):
            if satisfy(DICE_LIST[i1],DICE_LIST[i2]):
                total+=1
    return total

## Problème 91

In [17]:
class dot:
    def __init__(self,x,y):
        self.x=x
        self.y=y

    def __str__(self):
        return '({},{})'.format(self.x,self.y)

    def __add__(self,other):
        return dot(self.x+other.x,self.y+other.y)
    
    def __sub__(self,other):
        return dot(self.x-other.x,self.y-other.y)
    
    def __mul__(self,other):
        return self.x*other.x + self.y*other.y
    
    def origin():
        return dot(0,0)
    
    def orth(self):
        return dot(-self.y,self.x)
    
def orthogonal(U,V):
    return U*V==0

def rectinA(A,B,C):
    return orthogonal(B-A,C-A)

def direct(A,B,C):
    return (C-A)*(B-A).orth() > 0

def isrect(A,B,C):
    return rectinA(A,B,C) or rectinA(B,C,A) or rectinA(C,A,B)

def Pb91(M=50):
    O=dot.origin()
    counter=0
    for px in range(M+1):
        for py in range(M+1):
            for qx in range(M+1):
                for qy in range(M+1):
                    P=dot(px,py)
                    Q=dot(qx,qy)
                    if direct(O,P,Q) and isrect(O,P,Q):
                        counter+=1
    return counter

## Problème 92

In [18]:
from tqdm import tqdm

def Pb92(M=10**7):
    LIMIT_CHAIN={1:1,89:89}
    
    def successor(n):
        return sum([int(c)**2 for c in str(n)])
    
    def Pb92_aux(n):
        if n not in LIMIT_CHAIN:
            LIMIT_CHAIN[n]=Pb92_aux(successor(n))
        return LIMIT_CHAIN[n]
    
    for i in tqdm(range(2,M)):
        Pb92_aux(i)
    
    return len([i for i in range(1,M) if LIMIT_CHAIN[i]==89])

## Problème 93



By using each of the digits from the set, {1, 2, 3, 4}, exactly once, and making use of the four arithmetic operations (+, −, *, /) and brackets/parentheses, it is possible to form different positive integer targets.

For example,

```
8 = (4 * (1 + 3)) / 2
14 = 4 * (3 + 1 / 2)
19 = 4 * (2 + 3) − 1
36 = 3 * 4 * (2 + 1)
```

Note that concatenations of the digits, like 12 + 34, are not allowed.

Using the set, {1, 2, 3, 4}, it is possible to obtain thirty-one different target numbers of which 36 is the maximum, and each of the numbers 1 to 28 can be obtained before encountering the first non-expressible number.

Find the set of four distinct digits, a < b < c < d, for which the longest set of consecutive positive integers, 1 to n, can be obtained, giving your answer as a string: abcd.


In [19]:
def tree_gen():
    OP = ['+','-','*','/']
    SYMB = [['{}']]

    for i in range(1,4):
        SYMB.append([])
        for j in range(i):
            for L in SYMB[j]:
                for R in SYMB[i-j-1]:
                    for c in OP:
                        SYMB[i].append('({}{}{})'.format(L,c,R))
    return SYMB[3]

In [20]:
from itertools import permutations
from tqdm.notebook import tqdm_notebook
import scipy

S=tree_gen()

def values( L ):
    E=set()
    for c in S:
        for LP in permutations(L):
            try:
                n = eval(c.format(LP[0],LP[1],LP[2],LP[3]))
                if n==int(n) and n>0:
                    E.add(int(n))
            except ZeroDivisionError:
                pass
    return sorted(E)

def maxseq(E):
    curval=E[0]
    curlen=1
    maxlen=1
    for i in range(1,len(E)):
        if not E[i]-i==1:
            return i
    return len(E)

def Pb93():
    
    pbar=tqdm_notebook(total=int(scipy.special.binom(9,4)))
    
    maxlen=0
    curseq=''
    for a in range(1,10):
        for b in range(a+1,10):
            for c in range(b+1,10):
                for d in range(c+1,10):
                    
                    pbar.update()
                    
                    if ( curlen := maxseq( values([a,b,c,d]) ) ) > maxlen:
                            maxlen = curlen
                            curseq = '{}{}{}{}'.format(a,b,c,d)
    return curseq

## Problème 94

It is easily proved that no equilateral triangle exists with integral length sides and integral area. However, the almost equilateral triangle 5-5-6 has an area of 12 square units.

We shall define an almost equilateral triangle to be a triangle for which two sides are equal and the third differs by no more than one unit.

Find the sum of the perimeters of all almost equilateral triangles with integral side lengths and area and whose perimeters do not exceed one billion (1,000,000,000).

Probable solution : énumérer des triples pythagoriciens $(a,b,c)=2\times((n+e)/2,h,n)$ (avec $e=\pm1$), autrement dit $c=2n\equiv 0[2]$ pair, $|c-2a|=2|e|=2$, et $4\mathcal{A}=a\times b \equiv 0[4]$.

In [21]:
from tqdm.notebook import tqdm_notebook
from math import isqrt

def Pb94_slow(PER=10**9):
    
    SQList = set()
    for i in range(10**5):
        SQList.add(i**2)
    
    PERTOT=0
    
    for n in tqdm_notebook(range(2,PER)):
        for e in [-1,1]:
            N=(n+e)**2
            N=N*(4*n**2-N)
            
            if N%16==0:
                AA=N//16
                if AA in SQList and (3*n+e<=PER):
                    PERTOT+=3*n+e
    return PERTOT

def Pb94(PER=10**9):
    PERtot=0
    PERsq=isqrt((PER+1)//2)
    
    SQList = [i*i for i in range(PERsq+1)]

    for m in tqdm_notebook(range(2,PERsq+1)):
        
        if m%3==0:
            continue
        mm=m**2 # mm%3=1
        nn=(mm+2)//3
        
        if nn in SQList:
            n=isqrt(nn)
            k=1
        elif nn-1 in SQList:
            nn-=1
            n=isqrt(nn)
            k=2
        else:
            continue
            
        per=2*k*mm
        
        if per<=PER:
            PERtot+=per
        
    return PERtot

## Problème 95

The proper divisors of a number are all the divisors excluding the number itself. For example, the proper divisors of 28 are 1, 2, 4, 7, and 14. As the sum of these divisors is equal to 28, we call it a perfect number.

Interestingly the sum of the proper divisors of 220 is 284 and the sum of the proper divisors of 284 is 220, forming a chain of two numbers. For this reason, 220 and 284 are called an amicable pair.

Perhaps less well known are longer chains. For example, starting with 12496, we form a chain of five numbers:

12496 → 14288 → 15472 → 14536 → 14264 (→ 12496 → ...)

Since this chain returns to its starting point, it is called an amicable chain.

Find the smallest member of the longest amicable chain with no element exceeding one million.


In [22]:
from tqdm.notebook import tqdm_notebook as tqdm

def PROPDIVLIST(n):
    D=[[1] for _ in range(n+1)]
    D[0]=D[1]=[]
    
    for i in tqdm(range(2,n+1)):
            for k in range(2,n//i+1):
                D[k*i].append(i)
    return D

def Pb95(N=10**6):
    
    D=PROPDIVLIST(N)
   
    def iterate(n):
        return sum(D[n])
    
    SMALLESTCHAIN={}
    CHAINLENGTH={}
    
    smallest = 0
    maxlength = 0

    for i in tqdm(range(N+1)):
        L=[]
        head=i
        while (head not in SMALLESTCHAIN) and (head<=N) and (head not in L):
            L.append(head)
            head=iterate(head)

        if head>N:
            for j in L:
                SMALLESTCHAIN[j]=-1
                CHAINLENGTH[j]=-1
        elif head in SMALLESTCHAIN:
            for j in L:
                SMALLESTCHAIN[j]=SMALLESTCHAIN[head]
                CHAINLENGTH[j]=CHAINLENGTH[head]
        else:
            ind=L.index(head)
            CL=len(L)-ind
            SC=min(L[ind:])
            for j in L:
                SMALLESTCHAIN[j]=SC
                CHAINLENGTH[j]=CL
                
            if CL > maxlength:
                smallest=SC
                maxlength=CL
            elif CL==maxlength and SC<smallest:
                smallest=SC

    return smallest

## Problème 96

A well constructed Su Doku puzzle has a unique solution and can be solved by logic, although it may be necessary to employ "guess and test" methods in order to eliminate options (there is much contested opinion over this). The complexity of the search determines the difficulty of the puzzle; the example above is considered easy because it can be solved by straight forward direct deduction.

The 6K text file, sudoku.txt (right click and 'Save Link/Target As...'), contains fifty different Su Doku puzzles ranging in difficulty, but all with unique solutions.

By solving all fifty puzzles find the sum of the 3-digit numbers found in the top left corner of each solution grid; for example, 483 is the 3-digit number found in the top left corner of the solution grid above.

In [23]:
def options(SDQ,x,y):
    O=set(range(1,10))
    
    if SDQ[x][y] in O:
        return [SDQ[x][y]]

    u,v=x-x%3,y-y%3
    for i in range(3):
        for j in range(3):
            if SDQ[u+i][v+j] in O:
                O.remove(SDQ[u+i][v+j])

    for z in range(9):
        if SDQ[x][z] in O:
            O.remove(SDQ[x][z])
        if SDQ[z][y] in O:
            O.remove(SDQ[z][y])
            
    return sorted(O)

def fill_known_cells(SDQ):
    L={}
    
    changes = False

    for x in range(9):
        for y in range(9):
            if SDQ[x][y] == 0:
                L[(x,y)]=options(SDQ,x,y)

    for (x,y) in L:
        if len(L[(x,y)])==1:
            SDQ[x][y] = L[(x,y)][0]
            changes = True
            
    return changes

def zero_in(SDQ):
    return 0 in [min(L) for L in SDQ]

In [24]:
## Solver "optimisé" qui explose dès que le problème est un poil complexe...

def solver(G):
    from copy import deepcopy
    
    GG = deepcopy(G)
    while fill_known_cells(GG):
        None
        
    if zero_in(GG):
        O=range(1,10)
        u,v=0,0
        
        for x in range(9):
            for y in range(9):
                if SDQ[x][y] == 0:
                    L=options(SDQ,x,y)
                    if len(L)<=len(O):
                        O=L
                        u,v=x,y
        if len(O)==0:
            return False
        
        for o in O:
            GG[u][v] = o
            GGG=solve(GG)
            if GGG != False:
                return GGG        
        return False
    
    else:
        return GG

In [25]:
def solve_backtrack(G,n):
    
    if n==81:
        return True
    
    x,y=n%9,n//9
    
    if G[x][y] != 0:
        return solve_backtrack(G,n+1)
    else:
        O = options(G,x,y)
        if len(O)==0:
            return False
        else:
            for o in O:
                G[x][y]=o
                if solve_backtrack(G,n+1):
                    return True
            G[x][y]=0
            return False

In [26]:
from tqdm.notebook import tqdm_notebook as tqdm

def Pb96():

    import re

    def gridify(L):
        G = [[0 for _ in range(9)] for _ in range(9)]
        for n in range(81):
            G[n%9][n//9]=int(L[n])
        return G

    with open('files/Pb96.txt','r') as file:
        SDQList = re.split('Grid [\d][\d]\n',file.read())[1:]
        print(file.read())
        SDQList = [gridify(S.replace('\n','')) for S in SDQList]
    
    TOTAL=0
    
    for G in tqdm(SDQList):
        fill_known_cells(G)
        solve_backtrack(G,0)
        TOTAL+=10*(10*G[0][0]+G[1][0])+G[2][0]
    
    return TOTAL

## Problème 97

The first known prime found to exceed one million digits was discovered in 1999, and is a Mersenne prime of the form 2^6972593−1; it contains exactly 2,098,960 digits. Subsequently other Mersenne primes, of the form 2^p−1, have been found which contain more digits.

However, in 2004 there was found a massive non-Mersenne prime which contains 2,357,207 digits: 28433×2^7830457+1.

Find the last ten digits of this prime number.


In [27]:
def Pb97():
    return (28433*pow(2,7830457,10**10)+1)%10**10

## Problème 98

By replacing each of the letters in the word CARE with 1, 2, 9, and 6 respectively, we form a square number: 1296 = 36². What is remarkable is that, by using the same digital substitutions, the anagram, RACE, also forms a square number: 9216 = 96². We shall call CARE (and RACE) a square anagram word pair and specify further that leading zeroes are not permitted, neither may a different letter have the same digital value as another letter.

Using words.txt (right click and 'Save Link/Target As...'), a 16K text file containing nearly two-thousand common English words, find all the square anagram word pairs (a palindromic word is NOT considered to be an anagram of itself).

What is the largest square number formed by any member of such a pair?

NOTE: All anagrams formed must be contained in the given text file.

In [28]:
def numbers(pair):
    x,y = pair
    
    maxnum = 0
    
    for sq in SQUARES[len(x)] :
        
        stop = False
        
        num_map = {}
        values = set()
        for i in range(len(x)):
            if x[i] in num_map and num_map[x[i]] != int(str(sq)[i]) :
                stop = True
                break
            num_map[x[i]]=int(str(sq)[i])
            values.add(num_map[x[i]])
                
        if stop or len(set(x))!=len(values):
            continue
            
        y_num=0
        for c in y:
            y_num*=10
            y_num+=num_map[c]
            
        if y_num in SQUARES[len(x)]:
            maxnum = max(maxnum,sq,y_num)
            
    return maxnum

In [29]:
def Pb98():
    
    with open('files/Pb98.txt','r') as file:
        word_salad = file.read().replace('"','').split(',')

    anapairs = set()

    for i in range(len(word_salad)):
        for j in range(i):
            if sorted(word_salad[i]) == sorted(word_salad[j]):
                anapairs.add((word_salad[i],word_salad[j]))
                
    return max([numbers(pair) for pair in anapairs])

## Problème 99

In [30]:
from sympy import log

def Pb99():
    
    L = []
    
    with open('files/Pb99.txt','r') as file:
        pairlist = file.read().splitlines()
        for c in pairlist:
            i,j=c.split(',')
            L.append( log(int(i))*int(j) )
            
    return L.index(max(L))+1

## Problème 100



If a box contains twenty-one coloured discs, composed of fifteen blue discs and six red discs, and two discs were taken at random, it can be seen that the probability of taking two blue discs, P(BB) = (15/21)×(14/20) = 1/2.

The next such arrangement, for which there is exactly 50% chance of taking two blue discs at random, is a box containing eighty-five blue discs and thirty-five red discs.

By finding the first arrangement to contain over 10^12 = 1,000,000,000,000 discs in total, determine the number of blue discs that the box would contain.


On va chercher $u$ et $w=u+v$ tels que $\frac{u(u-1)}{w(w-1)}=\frac{1}{2}$.

In [31]:
from math import isqrt

def w_of_u(u):
    return isqrt(2*u*(u-1))+1

def valid(u,w):
    return w*(w-1)==2*u*(u-1)

In [32]:
def Pb100_slow(p=12):
    T=10**p
    S={}
    for w in range(int(T/(2**.5)),T):
        S[w*(w-1)]=w
    for w in range(T,10*T):
        ww=w*(w-1)
        if ww//2 in S:
            return S[ww//2]
        S[ww]=w

In [33]:
def Pb100(N=10**12):
    b,t=85,85+35
    
    while t<N:
        b,t=3*b+2*t-2,4*b+3*t-3
    
    return b,t

## Problème 101

If we are presented with the first k terms of a sequence it is impossible to say with certainty the value of the next term, as there are infinitely many polynomial functions that can model the sequence.

As an example, let us consider the sequence of cube numbers. This is defined by the generating function,
un = n^3: 1, 8, 27, 64, 125, 216, ...

Suppose we were only given the first two terms of this sequence. Working on the principle that "simple is best" we should assume a linear relationship and predict the next term to be 15 (common difference 7). Even if we were presented with the first three terms, by the same principle of simplicity, a quadratic relationship should be assumed.

We shall define OP(k, n) to be the nth term of the optimum polynomial generating function for the first k terms of a sequence. It should be clear that OP(k, n) will accurately generate the terms of the sequence for n ≤ k, and potentially the first incorrect term (FIT) will be OP(k, k+1); in which case we shall call it a bad OP (BOP).

As a basis, if we were only given the first term of sequence, it would be most sensible to assume constancy; that is, for n ≥ 2, OP(1, n) = u1.

Hence we obtain the following OPs for the cubic sequence:
```
OP(1, n) = 1 	1, *1, 1, 1, ...
OP(2, n) = 7n−6 	1, 8, *15, ...
OP(3, n) = 6n^2−11n+6      	1, 8, 27, *58, ...
OP(4, n) = n^3 	1, 8, 27, 64, 125, ...
```
Clearly no BOPs exist for k ≥ 4.

By considering the sum of FITs generated by the BOPs (indicated in red above), we obtain 1 + 15 + 58 = 74.

Consider the following tenth degree polynomial generating function:

$u_n = 1 − n + n^2 − n^3 + n^4 − n^5 + n^6 − n^7 + n^8 − n^9 + n^{10}$

Find the sum of FITs for the BOPs.


In [34]:
def Pb101():
    
    def f(x,n=10):
        if n==0:
            return 1
        else:
            return 1-x*f(x,n-1)

    VALS=[f(x) for x in range(1,11)]

    def evaluate(P,x):
        y=0
        for p in P[::-1]:
            y*=x
            y+=p
        return y

    import sympy
    from fractions import Fraction

    X=sympy.symbols('X')

    def lagrange (Lx, Ly):
        if  len(Lx)!= len(Ly):
            return 1
        y=0
        for k in range ( len(Lx) ):
            t=1
            for j in range ( len(Lx) ):
                if j != k:
                    t=t* ( (P-Lx[j]) /(Lx[k]-Lx[j]) )
            y+= t*Ly[k]
        return sympy.simplify(y)

    def interpolate(VALS,k):
        D = lagrange(range(1,k+1),VALS[:k]).as_coefficients_dict()
        return [Fraction(str(D[X**i])) for i in range(len(D))]

    PList = [interpolate(VALS,k) for k in range(1,11)]
    FIT = [evaluate(PList[k-1],k+1) for k in range(1,11)]
    return int(sum(FIT))

## Problème 102



Three distinct points are plotted at random on a Cartesian plane, for which -1000 ≤ x, y ≤ 1000, such that a triangle is formed.

Consider the following two triangles:

A(-340,495), B(-153,-910), C(835,-947)

X(-175,41), Y(-421,-714), Z(574,-645)

It can be verified that triangle ABC contains the origin, whereas triangle XYZ does not.

Using triangles.txt (right click and 'Save Link/Target As...'), a 27K text file containing the co-ordinates of one thousand "random" triangles, find the number of triangles for which the interior contains the origin.

NOTE: The first two examples in the file represent the triangles in the example given above.


In [35]:
class dot:
    def __init__(self,x,y):
        self.x=x
        self.y=y

    def __str__(self):
        return '({},{})'.format(self.x,self.y)

    def __add__(self,other):
        return dot(self.x+other.x,self.y+other.y)

    def __sub__(self,other):
        return dot(self.x-other.x,self.y-other.y)

    def __mul__(self,other):
        return self.x*other.x + self.y*other.y

    def origin():
        return dot(0,0)

    def orth(self):
        return dot(-self.y,self.x)

def zero_inside(A,B,C):
    s1 = (B-A).orth()*A
    s2 = (C-B).orth()*B
    s3 = (A-C).orth()*C
    return (s1 >= 0 and s2 >= 0 and s3 >= 0) or (s1 <= 0 and s2 <= 0 and s3 <= 0)

In [36]:
def Pb102():

    with open('files/Pb102.txt','r') as file:
        trilist = file.read().splitlines()

    total=0

    for triangle in trilist:
        ax,ay,bx,by,cx,cy=(int(c) for c in triangle.split(','))
        A=dot(ax,ay)
        B=dot(bx,by)
        C=dot(cx,cy)

        total += zero_inside(A,B,C)

    return total

## Problèmes 103, 105 et 106

$A\subset \mathbb{N}^*$ un ensemble (fini à $n$ éléments). On note $S(A)=\sum_{i\in A} i$. $A$ est spécial si :
- $\forall B,C\subset A$, $B,C\neq \emptyset,B\cap C=\emptyset, S(B)\neq S(C)$,
- si de plus $|B|>|C|$ alors $S(B)>S(C)$.

### 103

On pose $A_n^*$ l'ensemble spécial qui minimise $S(A)$ pour $|A|=n$ fixé. Notamment:
- $A_1^*=\{1\}$,
- $A_2^*=\{1,2\}$,
- $A_3^*=\{2,3,4\}$,
- $A_4^*=\{3,5,6,7\}$,
- $A_5^*=\{6,9,11,12,13\}$,
- $A_6^*=\{11, 18, 19, 20, 22, 25\}$.

Si $A=\{a_1,\dots,a_n\}$, et $b=a_{n//2+1}$, alors on peut définir $B(A)=\{b,a_1+b,\dots,a_n+b\}$.
Jusqu'à $n=4$ on a $A_{n+1}^*=B\left(A_n^*\right)$, mais pas au rang $n=5$. En effet, $B\left(A_5^*\right)=\{11, 17, 20, 22, 23, 24\}$ (avec $S(B) = 117$) est bien spécial mais pas optimal.

On peut associer à $A$ la set string, la concaténation de ses nombres. Ainsi, on peut associer `111819202225` à $A_6^*$. Quelle est la chaîne de $A_7^*$ ?

### 105

Le fichier `Pb105.txt` contient 100 listes (non-ordonnées) de nombres. Identifier les ensembles spéciaux $\left(A_i\right)$ dans la liste, et retourner $\sum_i S\left(A_i\right)$.

In [37]:
def B(A):
    n=len(A)
    b=A[n//2]
    return [b+a for a in [0]+A]

def S(A):
    return sum(A)

A=[11,18,19,20,22,25]
B=B(A)
B

[20, 31, 38, 39, 40, 42, 45]

In [38]:
import numpy as np

def Pb103(step=7,MAX=60):

    U=[]
    for i in range(1,MAX+1):
        for j in range(i+1,MAX+1):
            U.append([i,j])
    
    for _ in range(2,step):
        U=NEXT(U,MAX)

    s=''
    for i in U[ np.argmin([S(A) for A in U]) ] :
        s+=str(i)

    return s

In [39]:
from itertools import combinations

def SUBSETDIFF(A,k):
    
    for n in range(2,(len(A)+1)//2+1):
        for B in combinations(A,n):
            for C in combinations(set(A)-set(B),n-1):
                if S(B)==S(C)+k:
                    return False
    return True

def NEXT(U,MAX):
    V=[]
    n=len(U[0])
    for A in U:
        
        kmin=1
        imin=0
        imax=n-1
        while imin<imax:
            kmin+=A[imax]-A[imin]
            imax-=1
            imin+=1
        kmin=max(kmin,A[-1]+1)
        
        kmax=A[0]+A[1]
        imin=2
        imax=n-1
        while imin<imax:
            kmax+=A[imin]-A[imax]
            imax-=1
            imin+=1
        kmax=min(kmax,MAX+1)
        
        for k in range(kmin,kmax):
            if SUBSETDIFF(A,k):
                V.append(A+[k])

    return V

In [40]:
def SUBSETDIFF(A,k):
    for n in range(2,(len(A)+1)//2+1):
        for B in combinations(A,n):
            for C in combinations(set(A)-set(B),n-1):
                if S(B)==S(C)+k:
                    return False
    return True

def NO_EQUALITY(A):
    for k in range(2,len(A)):
        if not SUBSETDIFF(A[:k],A[k]):
            return False
    return True
    
def WELL_ORDERED(A):
    if len(A)<=1:
        return True
    elif len(A)==2:
        return A[0]<A[1]
    
    elif WELL_ORDERED(A[:-1]):
        n=len(A)-1
        
        kmin=1
        imin=0
        imax=n-1
        while imin<imax:
            kmin+=A[imax]-A[imin]
            imax-=1
            imin+=1
        kmin=max(kmin,A[-2]+1)
        
        kmax=A[0]+A[1]
        imin=2
        imax=n-1
        while imin<imax:
            kmax+=A[imin]-A[imax]
            imax-=1
            imin+=1
        
        return A[-1]>=kmin and A[-1]<kmax
    
    return False

def SPECIAL(A):
    B=sorted(A)
    return WELL_ORDERED(B) and NO_EQUALITY(B)

In [41]:
def Pb105():
    with open('files/Pb105.txt','r') as file:
        LISTS=[ [int(n) for n in C.split(',')] for C in file.read().split('\n') ]
    
    total = 0
    
    for A in LISTS:
        if SPECIAL(A):
            total+=S(A)

    return total

### 106

Si $n=|A|$ est fixé, on peut former $p(n)=\sum_{i=1}^n \sum_{j=1}^{\min(i,n-i)} \binom{n}{i,j}$ paires d'ensembles. Ainsi, $p(4)=25$. Cependant, on peut s'assurer que $A$ est spécial en nettement moins d'opérations.

In [42]:
def count_pairs():

    from scipy.special import binom

    def multinomial(params):
        if len(params) == 1:
            return 1
        return binom(sum(params), params[-1]) * multinomial(params[:-1])

    def subsetpairs(n):
        total=0
        for i in range(1,n):
            for j in range(1,min(i,n-i)+1):
                if i==j:
                    total += multinomial([i,j,n-i-j])/2
                else:
                    total += multinomial([i,j,n-i-j])
        return int(total)

    D={k:subsetpairs(k) for k in range(1,13)}
    print(D)
    
count_pairs()

{1: 0, 2: 1, 3: 6, 4: 25, 5: 90, 6: 301, 7: 966, 8: 3025, 9: 9330, 10: 28501, 11: 86526, 12: 261625}


Parmi les $25$ paires pour $n=4$, lorsque $A$ est ordonné, il suffit d'en tester une seule pour l'égalité $S(B)=S(C)$ pour déterminer si $A$ est spécial. Lorsque $n=7$, de façon analogue, il suffit d'en tester $70$ parmi les $966$ au total. Lorsque $n=12$, quel est le nombre minimal de tests nécessaires ?

Résultat partiel : `4:1, 5:5, 6:20, 7:70`...

In [43]:
from itertools import combinations

def PERMUT(A,B):
    if len(A)==0:
        return True
    else:
        return A[0]<B[0] and PERMUT(A[1:],B[1:])

def Pb106(N=12):
    total=0
    S=set(range(N))
    for n in range(2,N//2+1):
        for B in combinations(S,n):
            for C in combinations(S-set(B),n):
                if min(B)<min(C) and not PERMUT(sorted(B),sorted(C)):
                    total+=1
    return total

Pb106(12)

21384

## Problème 104

The Fibonacci sequence is defined by the recurrence relation:

    Fn = Fn−1 + Fn−2, where F1 = 1 and F2 = 1.

It turns out that F541, which contains 113 digits, is the first Fibonacci number for which the last nine digits are 1-9 pandigital (contain all the digits 1 to 9, but not necessarily in order). And F2749, which contains 575 digits, is the first Fibonacci number for which the first nine digits are 1-9 pandigital.

Given that Fk is the first Fibonacci number for which the first nine digits AND the last nine digits are 1-9 pandigital, find k.

In [44]:
from sympy import fibonacci

DIGITS = set([str(i) for i in range(1,10)])

def fst_pandigit(n):
    if len(str(n))<9:
        return False
    else:
        return set(str(n)[:9])==DIGITS

def lst_pandigit(n):
    if len(str(n))<9:
        return False
    else:
        return set(str(n)[-9:])==DIGITS

def Pb104():
    U,V=1,1
    k=2
    while True:
        if lst_pandigit(V) and fst_pandigit(fibonacci(k)):
            return k            
        U,V=V,(U+V)%10**10
        k+=1

## Problème 107

In [45]:
def minspantree(M):
    SIZE = np.shape(M)[0]
    Forest=[i for i in range(SIZE)]
    Edges={}
    for i in range(SIZE):
        for j in range(i+1,SIZE):
            if M[i,j] in Edges:
                Edges[M[i,j]].append((i,j))
            else:
                Edges[M[i,j]]=[(i,j)]
    if -1 in Edges:
        Edges.pop(-1)
    MINSPANTREE=[]
    for l in sorted(Edges):
        for (i,j) in Edges[l]:
            if Forest[i] != Forest[j]:
                MINSPANTREE.append((i,j))
                Fj=Forest[j]
                for k in range(SIZE):
                    if Forest[k]==Fj:
                        Forest[k]=Forest[i]
    return MINSPANTREE

In [46]:
def WEIGHT(TREE,M):
    return sum(M[i,j] for (i,j) in TREE)
def TOTAL(M):
    SIZE = np.shape(M)[0]
    total = 0
    for i in range(SIZE):
        for j in range(i+1,SIZE):
            if M[i,j]>0:
                total+=M[i,j]
    return total
def ANSWER(M):
    return TOTAL(M)-WEIGHT(minspantree(M),M)

In [47]:
def Pb107():
    with open('files/Pb107.txt','r') as file:
        M = np.matrix([ [int(n) for n in C.replace('-','-1').split(',')] for C in file.read().split('\n') ])
    return ANSWER(M)

## Problèmes 108 et 110

In [48]:
PRIME = [2,3,5,7,11,13,17,19,23,29,31,37,41,43]
T = 1000

def PRODPOW(PRIME,POW):
    p=1
    for i in range(len(POW)):
        p*= PRIME[i]**POW[i]
    return p

def DIVCOUNTSQ(S):
    p=1
    for v in S:
        p*=(2*v+1)
    return p

def test(S):
    return DIVCOUNTSQ(S)>2*T

def fix(S):
    Q=DIVCOUNTSQ(S[1:])
    R=(2*T//Q - 1)//2 + 1
    S[0]=R
    return None

def NEXT(S):
    change = False
    for i in range(0,len(S)-1):
        if S[i]>S[i+1]:
            p=S[i+1]+1
            imax=i+1
            change=True
            break
    if change:
        for i in range(0,imax+1):
            S[i]=p
        fix(S)
    return change

In [49]:
T=1000
def Pb108():
    n = 10**8
    for l in range(1,11):
        S = [1 for _ in range(l)]
        fix(S)
        n=min(n,PRODPOW(PRIME,S))
        while NEXT(S):
            n=min(n,PRODPOW(PRIME,S))
    return n

In [50]:
T=4*10**6
def Pb110():
    n = PRODPOW(PRIME,[1 for _ in PRIME])
    for l in range(1,len(PRIME)+1):
        S = [1 for _ in range(l)]
        fix(S)
        n=min(n,PRODPOW(PRIME,S))
        while NEXT(S):
            n=min(n,PRODPOW(PRIME,S))
    return n

## Problème 109

In [51]:
def value(dart):
    if dart=='':
        return 0
    elif dart[0]=='S':
        return int(dart[1:])
    elif dart[0]=='D':
        return 2*int(dart[1:])
    elif dart[0]=='T':
        return 3*int(dart[1:])
    
def score(a,b,c):
    return value(a)+value(b)+value(c)

def Pb109(SCOREMAX=99):
    LAST=['D{}'.format(i) for i in range(1,21)]
    LAST.append('D25')

    ALL=['']+['S{}'.format(i) for i in range(1,21)]+['S25']+LAST+['T{}'.format(i) for i in range(1,21)]

    COMBI=[set() for i in range(171)]

    for D1 in ALL:
        for D2 in ALL:
            for D3 in LAST:
                COMBI[score(D1,D2,D3)].add((min(D1,D2),max(D1,D2),D3))
                
    return sum(len(S) for S in COMBI[:SCOREMAX+1])

## Problème 111

In [52]:
from sympy import isprime
from itertools import combinations

class ENUMPOS:
    def __init__(self,C,N,d):
        self.C=C
        self.d=d
        self.N=N
        self.index= 10**(len(C)-1) if (0 in C) else 0
        while str(self.d) in str(self.index):
            self.index += 1
        self.maxindex = 10**len(C)
        
    def __iter__(self):
        return self
    
    def __next__(self):
        if self.index < self.maxindex :
            item = [str(self.d) for _ in range(self.N)]
            fill = str(self.index).rjust(len(self.C),'0')
            for i in range(len(self.C)):
                item[self.C[i]]=int(fill[i])
            self.index += 1
            while str(self.d) in str(self.index):
                self.index += 1
            return int(''.join(str(c) for c in item))
        else:
            raise StopIteration

def is_prime(L):
    return isprime(int(''.join([str(i) for i in L])))

def BigD(N=4,d=1):
    BigSet=set()
    maxrep=0
    for k in range(0,N):
        for C in combinations(range(N),k):
            if d==0 and (0 not in C):
                continue           
            for n in ENUMPOS(C,N,d):
                if isprime(n):
                    maxrep=max(maxrep,C[0])
                    for j in range(k-1):
                        maxrep=max(maxrep,C[j+1]-C[j]-1)
                    maxrep=max(maxrep,k-C[-1])
                    BigSet.add(n)
        if len(BigSet)>0:
            return sum(BigSet)#,maxrep,len(BigSet)
    return 0

def Pb111(N=10):
    return sum(BigD(N,d) for d in range(10))

## Problème 112

In [53]:
def const(n):
    return len(set(list(str(n))))==1
    
def incr(n): # Vérifie la croissance
    return sorted(str(n))==list(str(n))

def decr(n):
    return sorted(str(n))==list(str(n))[::-1]

def zigzag(n):
    return (not incr(n)) and (not decr(n))

def Pb112():
    zzs=0
    n=100
    while 100*zzs < 99*n:
        n+=1
        zzs+=zigzag(n)
    return n

## Problème 113

In [54]:
from scipy.special import comb as binom

def num_incr(rlen=2): # Compte les croissants de longueur au plus rlen
    return int(binom(8+rlen,8,exact=True))

def num_decr(rlen=2): # Compte les croissants de longueur au plus rlen
    return sum(int(binom(first+rlen-1,first,exact=True)) for first in range(1,10))

def Pb113(power=10):
    return sum(num_incr(i)+num_decr(i)-9 for i in range(1,power+1))

## Problème 114

In [55]:
def Pb114(m=3,n=50):
    F=[1 for _ in range(n+1)]
    for i in range(m,n+1):
        F[i] = (i-m+1) + 1 + sum((p-m+1)*F[i-p-1] for p in range(m,i))
    return F[-1]

## Problème 115

In [56]:
def Pb115(m=50):
    F=[1 for _ in range(m)]
    i=m-1
    while F[-1] <= 10**6:
        i+=1
        F.append( (i-m+1) + 1 + sum((p-m+1)*F[i-p-1] for p in range(m,i)) )
    return i

## Problème 116

In [57]:
def Pb116_aux(m=3,n=50):
    F=[1 for _ in range(m)]
    F.append(2)
    for i in range(m+1,n+1):
        F.append( 1 + sum(F[i-p] for p in range(m,i+1)) )
    return F[-1]-1

def Pb116(n=50):
    return Pb116_aux(2,n)+Pb116_aux(3,n)+Pb116_aux(4,n)

## Problème 117

In [58]:
def Pb117(n=50):
    F=[1, 1, 2, 4]
    i=len(F)-1
    while i<n:
        i+=1
        F.append( F[i-1]+F[i-2]+F[i-3]+F[i-4] )
    return F[-1]

## Problème 118

In [59]:
from sympy import isprime
from itertools import permutations
from sympy.utilities.iterables import multiset_partitions

[p for p in multiset_partitions([1, 1, 1, 2], 2)]
[[[1, 1, 1], [2]], [[1, 1, 2], [1]], [[1, 1], [1, 2]]]

def conc(q):
    return int(''.join(str(d) for d in q))

def Pb118():
    PRIMES = {}
    
    def Pb118_aux(partition):
        if len(partition)==0:
            return 1
        else:
            p = partition[0]
            total = 0
            for q in permutations(p):
                n = conc(q)
                if n not in PRIMES:
                    PRIMES[n] = isprime(n)
                if PRIMES[n]:
                    total += Pb118_aux(partition[1:])
            return total
        
    TOTAL = 0
                
    for k in range(2,7):
        for partition in multiset_partitions(range(1,10),k):
            partition.sort(key=len)
            TOTAL += Pb118_aux(partition)
            
    return TOTAL

## Problème 119

In [60]:
def sumdigits(n):
    return sum(int(d) for d in str(n))

def Pb119(N=2):
    S=set()
    for n in range(2,10**N):
        m=n
        for p in range(2,10**N):
            m*=n
            if sumdigits(m)==n:
                S.add(m)
    return sorted(S)[29]

## Problème 120

Let r be the remainder when (a−1)^n + (a+1)^n is divided by a^2.

For example, if a = 7 and n = 3, then r = 42: 6^3 + 8^3 = 728 ≡ 42 mod 49. And as n varies, so too will r, but for a = 7 it turns out that rmax = 42.

For 3 ≤ a ≤ 1000, find ∑ rmax.

In [61]:
from tqdm.notebook import tqdm_notebook as tqdm

def powmod(a,n):
    aa=a*a
    return (pow(a-1,n,aa)+pow(a+1,n,aa))%aa

def rmax(a):
    
    rmax=0    
    b,c=a-1,a+1
    aa,bb,cc=a*a,1,1
    
    for _ in range(a*a):
        bb*=b
        bb%=aa
        cc*=c
        cc%=aa
        if (bb+cc)%aa>rmax:
            rmax=(bb+cc)%aa
    
    return rmax    
    return max(powmod(a,n) for n in range(1,a*a))

def Pb120():
    return sum(rmax(a) for a in tqdm(range(3,1001)))

## Problème 121

A bag contains one red disc and one blue disc. In a game of chance a player takes a disc at random and its colour is noted. After each turn the disc is returned to the bag, an extra red disc is added, and another disc is taken at random.

Tour n: proba 1/(n+1) d'avoir un bleu.

The player pays £1 to play and wins if they have taken more blue discs than red discs at the end of the game.

If the game is played for four turns, the probability of a player winning is exactly 11/120, and so the maximum prize fund the banker should allocate for winning in this game would be £10 before they would expect to incur a loss. Note that any payout will be a whole number of pounds and also includes the original £1 paid to play the game, so in the example given the player actually wins £9.

Find the maximum prize fund that should be allocated to a single game in which fifteen turns are played.

In [62]:
from fractions import Fraction

def pwin_aux(L,k):
    if k>len(L):
        return Fraction(0)
    elif k==0:
        return Fraction(1)
    else:
        return L[0]*pwin_aux(L[1:],k-1) + (1-L[0])*pwin_aux(L[1:],k)
    
def pwin_comp(turns):
    L=[Fraction(1,n) for n in range(2,turns+2)]
    return pwin_aux(L,(turns+2)//2)

def Pb121(turns):
    pwin = pwin_comp(turns)
    return int(1/pwin)

## Problème 122

The most naive way of computing n15 requires fourteen multiplications:

    n × n × ... × n = n15

But using a "binary" method you can compute it in six multiplications:

    n × n = n2
    n2 × n2 = n4
    n4 × n4 = n8
    n8 × n4 = n12
    n12 × n2 = n14
    n14 × n = n15

However it is yet possible to compute it in only five multiplications:

    n × n = n2
    n2 × n = n3
    n3 × n3 = n6
    n6 × n6 = n12
    n12 × n3 = n15

We shall define m(k) to be the minimum number of multiplications to compute nk; for example m(15) = 5.

For 1 ≤ k ≤ 200, find ∑ m(k).


In [63]:
def NEXT(SL):
    L = []
    for S in SL:
        for i in range(len(S)):
            for j in range(len(S)-1,i-1,-1):
                if S[i]+S[j]>max(S) :
                    if S[i]+S[j]<=200 :
                        L.append( tuple(list(S)+[S[i]+S[j]]) )
                else:
                    break
    return L

def Pb122(STEP=11):
    SL=[[1]]
    DIC={1:0}
    i=0
    for _ in range(STEP):
        SL=NEXT(SL)
        i+=1
        for S in SL:
            if S[-1] not in DIC:
                DIC[S[-1]]=i
            
    return sum(DIC[d] for d in DIC),set(range(1,201))-set(DIC)

In [64]:
# Pb122(STEP=10) : (1571, {191})
# Je conjecture que DIC[191]=11.

## Problème 123

In [65]:
from sympy import primerange

def Pb123():
    PRIME = list(primerange(10**6))

    def powprime(PRIME,n):
        return ( pow(PRIME[n-1]-1,n,PRIME[n-1]**2)+pow(PRIME[n-1]+1,n,PRIME[n-1]**2) ) % PRIME[n-1]**2
    n=1
    while powprime(PRIME,n) < 10**10:
        n+=2
    
    return n

## Problème 124

In [66]:
def PRIMEDIVLIST(n=10**5):
    L = [ [] for _ in range(n+1) ]
    for i in range(2,n+1):
        if L[i]==[]:
            for j in range(1,n//i+1):
                L[i*j].append(i)
    return L

def PROD(L):
    if len(L)==0:
        return 1
    return L[0]*PROD(L[1:])

def Pb124(N=100000):
    L = PRIMEDIVLIST(N)
    E = list(range(1,N+1))
    E.sort(key = lambda n : PROD(L[n]))
    return E[10000-1]
    

## Problème 125

The palindromic number 595 is interesting because it can be written as the sum of consecutive squares: 6^2 + 7^2 + 8^2 + 9^2 + 10^2 + 11^2 + 12^2.

There are exactly eleven palindromes below one-thousand that can be written as consecutive square sums, and the sum of these palindromes is 4164. Note that 1 = 0^2 + 1^2 has not been included as this problem is concerned with the squares of positive integers.

Find the sum of all the numbers less than 10^8 that are both palindromic and can be written as the sum of consecutive squares.

In [67]:
import numpy as np
from math import isqrt

def Pb125(MAXVAL=10**8):

    SIZE=isqrt(MAXVAL+1)+1

    L=np.zeros((SIZE,SIZE),dtype=np.int64)
    PAL=np.zeros((SIZE,SIZE),dtype=bool)
    OVERFLOW=np.zeros((SIZE,SIZE),dtype=bool)
    
    for i in range(SIZE):
        L[i,i]=(i+1)**2
    for d in range(1,SIZE):
        for i in range(SIZE-d):
            l = L[i,i+d-1]+L[i+d,i+d]
            if OVERFLOW[i,i+d-1] or l>MAXVAL:
                OVERFLOW[i,i+d] = True
                break
            L[i,i+d]=l
            PAL[i,i+d] = (str(l)==str(l)[::-1])
    return sum(set(L[PAL]))